src/main/java/com/urbanairship/datacube/dbharnesses/HbaseBatchIncrementer.java

```diff
@@ -0,0 +1,248 @@
+     * (include partial failures where only a part of the batch succeeded, and an exception was thrown) and updates
+     * the success map and list with the results. It then rethrows whatever exception resulted from the HBase
+     * interaction.
+     *
+     * @param batchMap            The write operations to accomplish
+     * @param successfulAddresses The datacube address whose writes succeeded
+     * @param successfulRows      A map from HBase row key to the value in the row after completion of the increment
+     *                            operation
+     *
+     * @throws IOException
+     * @throws InterruptedException
+     */
+
+    public void batchIncrement(Map<Address, T> batchMap, List<Address> successfulAddresses, Map<byte[], byte[]> successfulRows) throws IOException, InterruptedException {
+        Iterable<List<Map.Entry<Address, T>>> partitions = Iterables.partition(batchMap.entrySet(), configuration.batchSize);
+
+        CubeIncrementResults.Builder cubeIncrementResultsBuilder = CubeIncrementResults.newBuilder();
+
+        int batchesPerFlush = 0;
+        Map<BoxedByteArray, Address> backwards = new HashMap<>();
+        for (List<Map.Entry<Address, T>> entries : partitions) {
+            Map<BoxedByteArray, T> increments = new HashMap<>();
+            batchesPerFlush++;
+            final long nanoTimeBeforeWrite = System.nanoTime();
+            for (Map.Entry<Address, T> entry : entries) {
+                final Address address = entry.getKey();
+                final BoxedByteArray rowKey = serializer.apply(address);
+
+                increments.put(rowKey, entry.getValue());
+                backwards.put(rowKey, address);
+            }
+
+            cubeIncrementResultsBuilder.add(this.hbaseIncrement(increments));
+
+            long writeDurationNanos = System.nanoTime() - nanoTimeBeforeWrite;
+            incrementerMetrics.batchWritesTimer.update(writeDurationNanos, TimeUnit.NANOSECONDS);
+
+            if (cubeIncrementResultsBuilder.ioException != null) {
+                break;
+            }
+
+            if (cubeIncrementResultsBuilder.interrupt != null) {
+                // still gotto report our successes
+                break;
+            }
+        }
+        incrementerMetrics.batchesPerFlushHisto.update(batchesPerFlush);
+
+        for (Map.Entry<BoxedByteArray, byte[]> entry : cubeIncrementResultsBuilder.successes.entrySet()) {
+            successfulAddresses.add(backwards.get(entry.getKey()));
+            successfulRows.put(entry.getKey().bytes, entry.getValue());
+        }
+
+        int failures = batchMap.size() - successfulAddresses.size();
+
+        if (failures > 0 || cubeIncrementResultsBuilder.ioException != null || cubeIncrementResultsBuilder.interrupt != null) {
+            incrementerMetrics.incrementFailuresPerFlush.update(failures);
+
+            if (cubeIncrementResultsBuilder.ioException != null) {
+                throw cubeIncrementResultsBuilder.ioException;
```

Comment:
Should wrap these in a new exception, otherwise you'll blow away any existing trace data.

Reply:
I chose not to wrap the io exception because doing so would require callers to understand that the CAUSE might be an instance of `RetriesExhaustedWithDetailsException` which seemed sort of distasteful to me. I thought it better to let the exception bubble direct from hbase. Additionally this matches the behavior of the hbase db harness which throws the unwrapped exception: https://github.com/urbanairship/datacube/blob/master/src/main/java/com/urbanairship/datacube/dbharnesses/HBaseDbHarness.java#L487

I could just as well not catch the exception at all, and accomplish all the logic needed to mark successes in a finally block. would that be preferable?

Reply:
Doing it all in finally blocks gets rull weird.

Reply:
Actually it's not that bad. This PR shows an implementation that uses finally instead of passing around an object: https://github.com/urbanairship/datacube/pull/96